block/src/Pos/DB/Block.hs

```diff
@@ -97,15 +97,40 @@ getTipBlock = getTipSomething "block" getBlock
 getSerializedBlock
     :: forall ctx m. (MonadRealDB ctx m)
     => HeaderHash -> m (Maybe ByteString)
-getSerializedBlock hh = do
-    DB{..} <- getBlockDataDB
-    liftIO $ Rocks.get rocksDB rocksReadOpts (blockKeyMod . sformat hashHexF $ hh)
+getSerializedBlock = getSerializedBlockOrUndo fst
 
 -- Get serialization of an undo data for block with given hash from Block DB.
 getSerializedUndo :: (MonadRealDB ctx m) => HeaderHash -> m (Maybe ByteString)
-getSerializedUndo hh = do
-    DB{..} <- getBlockDataDB
-    liftIO $ Rocks.get rocksDB rocksReadOpts (undoKeyMod . sformat hashHexF $ hh)
+getSerializedUndo = getSerializedBlockOrUndo snd
+
+getSerializedBlockOrUndo ::
+       MonadRealDB ctx m
+    => ((ByteString, ByteString) -> ByteString)
+    -> HeaderHash
+    -> m (Maybe ByteString)
+getSerializedBlockOrUndo blockOrUndoSelector hh = do
+    DB {..} <- getBlockDataDB
+    let rocksGet = Rocks.get rocksDB rocksReadOpts
+    let rocksPut = Rocks.put rocksDB rocksWriteOpts
+    let rocksDel = Rocks.delete rocksDB rocksWriteOpts
+    liftIO (rocksGet (blundKeyMod hh)) >>= \case
+        Nothing -> fmap blockOrUndoSelector <$>
+            consolidateBlund rocksGet rocksPut rocksDel
+        Just ser ->
+            eitherToThrow $
+            bimap DBMalformed (Just . blockOrUndoSelector) $
+            decodeFull' @(ByteString, ByteString) ser
+  where
+    consolidateBlund rocksGet rocksPut rocksDel = liftIO $ do
+        serBlock <- rocksGet (blockKeyMod hh)
+        serUndo <- rocksGet (undoKeyMod hh)
+        case (,) <$> serBlock <*> serUndo of
+            Just blund -> do
+                rocksPut (blundKeyMod hh) $ serialize' blund
+                rocksDel (blockKeyMod hh)
+                rocksDel (undoKeyMod hh)
+                return $ Just blund
+            Nothing -> return Nothing
```

Comment:
Is it okay to do nothing if we only found undo and not the block or vice-versa? Wouldn't it be better to delete them?

Reply:
I think that ideally we should delete them, but I don't think it will happen to anybody in practice (there are not many users) and I want to keep this code quite similar to IOHK's code, so I'd like to leave it as is.